AD-HOC feat (kolide): add kolide fleet

andrewhowdencom opened the pull request:
This will probably need a little work. At this time, 

> https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements

doesn't pass. In particular, I'm not quite sure how to set a persistent mysql password for the parent chart that propagates to the child chart.

**What this PR does / why we need it**:

This PR creates a kolide/fleet chart. Kolide's fleet tool is usefulf or querying a large chunk of hosts for specific requirements.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Nah this is random.

**Special notes for your reviewer**:

This is chunked in this was as I was writing a tutorial. See  https://medium.com/@andrewhowdencom/deploying-on-kubernetes-1-the-use-case-b4c201e2f049

Commit:
AD-HOC feat (kolide/fleet): Initial specification of the kolide/fleet chart

This is a super early commit of the kolide/fleet chart. The chart
scaffold itself is only partially implemented, as it's being documented
heavily as part of a blog series during it's construction and writing
has a necessarily high overhead.

Future work will build on this by adding dependencies, moving content
from `.to-do` to `templates` (or deleting it as required) etc.

== Design Notes ==

`.to-do` - The todo folder is essentially scrap, but has been committed
for transparency while undertaking the blog series.

Commit:
AD-HOC feat (requirements): Document the requirement of MySQL, Redis

The fleet application uses MySQL and Redis as it's persistent storage
layers. This commit adds the latest version of both at the time of writing,
such that as the application is designed it can be tested against
these installations.

Notes:

https://medium.com/@andrewhowdencom/deploying-on-kubernetes-3-dependencies-e024741a15df

Commit:
AD-HOC feat (Deployment): Add partially completed deployment

The deployment object is the primary object that is used to manage the
deloyed software. This commit adds one as it is completed thusfar from
the `.to-do` folder.

It is added half-completed such that the history can show significant
changes, and log in the git commits the justificaiton for them.

Commit:
AD-HOC refactor (Deployment): Change update strategy to recreate

Though the kolide application does not store any state directly attached
to it's canonical service (instead apparently using MySQL and Redis as
the stateful services), it's unclear that it will be able to manage
state transitions in which some replicas may have new configuration or
an application version, whereas others do not.

Accordingly, the author feels it is safer to simply kill all replicas
before rolling out a new version, such that the new version can assume
te same state across the deployment where replicated multiple times.

Commit:
AD-HOC feat (deployment): Fill out the RestartPolicy object with "Always"

There are various reasons that software will fail in a production
environment:

- OOM
- SegFault
- Application error
- Friday

The RestartPolicy denotes what happens in these circumstances. In this
case, the software being run is the `kolide/fleet` software, which
expects to be daemonized by any of the traditional system daemons.

Accordingly, following the pattern established by these daemons, the
policy of "Always Restart" has been implemented.

== Design Notes ==

=== Commit Order ===

This commit does not follow the pattern of the series that uses these
commits to explain the deployment. It is useful to make the deloyment
deployable, such that the author can iterate while writing.

Commit:
AD-HOC feat (configuration): Make fleet largely configurable

This commit introduces the fleet configuration property by way of a
configmap, as well as making most of the fleet configuration editable.

Some configuration was deliberately not managed as part of this
configmap including mysql passwords, etc. Additionally, the
configuration hasn't yet been documented.

Commit:
AD-HOC feat (Kolide): Configure the MySQL, Redis secrets

In previous work the majoritiy of the configuration for Kolide was
implemented. However, there was certain configuration that was
inherently sensitive in nature, and should have been stored in a secret
object.

This commit adds the stubs for that configuration. Unfortunately adding
that configuration itself is difficult, due to the way helm handles the
generation of secrets -- i.e. the reexecution of template functions with
each release.

Accordingly, this commit opts for deferring the management of the secret
information to the user.

== Notes ==

A suitable values.secret.yml (or appropriate encrypted file) is
generated through the following command:

  $ cat <<EOF > values.secret.yaml
  ---
  mysql:
    mysqlUser: "kolide"
    mysqlPassword: $(pwgen 32 1)
  redis:
    redisPassword: $(pwgen 32 1)
  fleet:
    auth:
      jwt_key: $(pwgen 32 1)
  EOF

Future work should probably put that in a README.

Commit:
AD-HOC feat (Kolide/Fleet): Add an installation job

The kolide/fleet application requires some installation prior to booting
the application. Additionally, the installation requires invocation of
the CLI client, and cannot be deferred to some sort of GUI.

This commit adds a post install and post upgrade script to invoke the DB
migration process. It appears idempotent, and upgrades that are not
associated with the change of an image should not cause any issues.

Commit:
AD-HOC feat (TLS): Document the TLS requirement, mount into container

In order to allow connection to fleet, fleet expects TLS configuration
(or TLS to be disabled). This commit introduces the required TLS
configuration, though does not generate the certificates.

Unfortunately, generating the certificates proved to expensive in terms
of time while the author was completing this work. Future work will
attempt to generate TLS certificates automatically.

Additionally, future work should allow disabling TLS entirely.

Commit:
AD-HOC feat (Service, Deployment, Secret): Add service, fix Redis

This commit adds a service such that the deployment is discoverable. It
uses the service defined by the starter template, but adjusts the
service to reflect the ingress on port 443 rather than 8080 so that
applications that connect via HTTPS can connect on the default port.

Additionally, it was caught during testing that the kolide application
does not have access to Redis. This was traced to a failure to mount the
secret into the container with the previous secret implementation.

Oops.

andrewhowdencom committed:
AD-HOC feat (kolide-fleet): Add liveness checking to the deployment

In previous work a deployment was crafted which deploys the kolide
application. However, there was no health checking implemented. Health
checking is required to ensure the application is up and functioning
correctly, and has not somehow frozen. Kolide can get into this state
when it is unable to connect to the upstream MySQL server or prior to
the application being installed, when the install job has not yet
completed.

andrewhowdencom committed:
AD-HOC feat (Annotations); Introduce customisable annotations

This commit introduces the ability to customise annotations on both the
service and the deployment specification. This should allow integration
of this chart with various third party tools, such as Prometheus.

Only the service and pod were implemented at this time. Further
extension is not planned, but should be simple to implement.

andrewhowdencom committed:
AD-HOC refactor (Resource Constraints): Drop the resource constraints

This commit modifies the resource constraints. The constraints were
evaluated against a low traffic cluster, and seem reasonable.

andrewhowdencom committed:
AD-HOC chore (kolide-fleet): Remove boilerplate

andrewhowdencom committed:
AD-HOC chore (Docs): Add limited documentation

andrewhowdencom committed:
AD-HOC fix (Deployment): Remove unnecessary privileged flag

andrewhowdencom committed:
AD-HOC chore (NOTES.txt): Remove reference to automated TLS generation

andrewhowdencom committed:
AD-HOC fix (kolide-fleet)): Fix various hard-coded references

andrewhowdencom committed:
AD-HOC chore (Lints): Fix lint failures

andrewhowdencom commented:
/assign @prydonius

unguiculus commented on stable/kolide-fleet/LICENSE:
```diff
@@ -0,0 +1,19 @@
+The MIT License (MIT)
```
Remove this file. This repo is Apache license.

unguiculus commented on stable/kolide-fleet/MAINTAINERS.md:
```diff
@@ -0,0 +1,4 @@
+The following document lists the maintainers of the project, as well as the username for each.
```
Remove this file. Maintainers are listed in Chart.yaml. Plus, you may add an OWNERS file.

unguiculus reviewed (changes requested):
Please read up on our guidelines. Make sure standard labels are correctly applied to all resources.

Please compare with what you get from `helm create` using Helm 2.8+ nd update accordingly.

unguiculus commented:
/assign

prydonius commented:
@andrewhowdencom if you think it'll need more work to meet the technical requirements, you could submit it to incubator for now?

andrewhowdencom committed:
AD-HOC refactor (incubator): Move to incubator spec

andrewhowdencom commented:
Hola @prydonius; 

Done. I intend to address the feedback; I just haven't had the budget yet. ^^

andrewhowdencom committed:
AD-HOC docs (*): Remove license, maintainers. Add Owners.

andrewhowdencom committed:
AD-HOC refactor (kolide-fleet): Bump labels to be in line with recent specs

andrewhowdencom committed:
AD-HOC style (Values.yaml): Fix style in values.yaml

andrewhowdencom commented:
Hola @unguiculus ;

I've implemented the requested changes. I'm still a little fuzzy about how to handle the MySQL stuff. 

The problem is that I need to set the MySQL password for two distinct values at the same time, or hoist the value from the MySQL chart to fleet. 

Do you have any preference as to which one it should be? In the case it's hoisted, I can sort of rely on MySQL to keep it consistent through releases (which I'm not sure if this chart does -- at any rate, it's that charts' problem). 

My initial plan of "mark it as an install job" doesn't work, as I can't override the MySQL secret. 

I might see if I can hoist it for now. :thinking:

andrewhowdencom commented:
Also PKI which is a whole other thing :sweat:

unguiculus commented:
Please merge in latest master.

You can specify custom values for CI:
https://github.com/kubernetes/charts/blob/master/test/README.md#providing-custom-test-values

andrewhowdencom committed:
AD-HOC feat (incubator/kolide-fleet): Add CI values

There are some values that are required to bootstrap this application by
default. This commit supplies these values for the CI such that CI can
pass.

See https://github.com/kubernetes/charts/blob/master/test/README.md#providing-custom-test-values

unguiculus commented on incubator/kolide-fleet/README.md:
```diff
@@ -0,0 +1,112 @@
+
+## Prerequisites
+
+- Kubernetes 1.3+ with Beta APIs enabled 
```
1.9+

andrewhowdencom replied:
:heavy_check_mark:

unguiculus commented on incubator/kolide-fleet/templates/deployment.yaml:
```diff
@@ -0,0 +1,99 @@
+---
+apiVersion: "extensions/v1beta1"
```
apps/v1

andrewhowdencom replied:
:heavy_check_mark:

unguiculus commented on incubator/kolide-fleet/templates/deployment.yaml:
```diff
@@ -0,0 +1,99 @@
+{{- if .Values.pod.annotations }}
+{{ toYaml .Values.pod.annotations | indent 8 }}
+{{ end }}
+        checksum.helm.kubernetes.io/secret: {{ include (print $.Chart.Name "/templates/secret.yaml") . | sha256sum }}
```
You should do this:

`{{ include (print .Template.BasePath "/secret.yaml") . | sha256sum }}`

andrewhowdencom replied:
:heavy_check_mark:

unguiculus commented on incubator/kolide-fleet/templates/deployment.yaml:
```diff
@@ -0,0 +1,99 @@
+    heritage: {{ .Release.Service }}
+  name: {{ template "fleet.fullname" . }}
+spec:
+  replicas: {{ default 2 .Values.deployment.replicas }}
```
Don't use the default function for simple default values. Defaults should be in `values.yaml`.

andrewhowdencom replied:
I'm reading this as "anything that can be defined in values.yaml should be", in which case :heavy_check_mark:

unguiculus commented on incubator/kolide-fleet/templates/service.yaml:
```diff
@@ -0,0 +1,34 @@
+apiVersion: "v1"
+metadata:
+  annotations:
+{{- if .Values.pod.annotations }}
```
.Values.service.annotations

andrewhowdencom replied:
:heavy_check_mark:

andrewhowdencom replied:
:heavy_check_mark:

unguiculus commented on incubator/kolide-fleet/templates/service.yaml:
```diff
@@ -0,0 +1,34 @@
+{{- if .Values.pod.annotations }}
+{{ toYaml .Values.service.annotations | indent 4 }}
+{{ end }}
+    {{ if eq .Values.service.type "LoadBalancer" }}service.beta.kubernetes.io/external-traffic: "OnlyLocal"{{ end }}
```
This if statement is unnecessary. The annnotation can be added with `service.annotations`. However, the annotation is obsolete with the field `externalTrafficPolicy` on the service spec. Add that and make it configurable.

andrewhowdencom replied:
:heavy_check_mark:

unguiculus commented on incubator/kolide-fleet/values.yaml:
```diff
@@ -0,0 +1,83 @@
+deployment:
+  replicas: 1
+
+resources:
```
Don't add defaults for resources. Comment them out.

andrewhowdencom replied:
:heavy_check_mark:

unguiculus commented on incubator/kolide-fleet/values.yaml:
```diff
@@ -0,0 +1,83 @@
+    json: true
+    ## Whether to disable the banner as part of the logs
+    disable_banner: true
+
```
Nesting by `mysql`, `redis`, and `fleet` makes sense. Move `pod`, `service`, and `resources` under `fleet`. The `image` and `replicas` should be directly under `fleet` then.

andrewhowdencom replied:
This one makes a little less sense to me. I will implement it as i've literally understood it, but I feel like `image` should be a property of `pod` and not directly. So, 

`fleet.pod.image` 

and not

`fleet.image`

whereas I understand you to be asking for the latter. I'll implement it as I understand in case I'm missing something, but I wanted to raise it.

unguiculus commented on incubator/kolide-fleet/values.yaml:
```diff
@@ -0,0 +1,83 @@
+    # prometheus.io/scheme: https
+    # prometheus.io/scrape: "true"
+  fleet:
+    image: kolide/fleet
```
```yaml
image:
  repository:
  tag:
  pullPolicy:
```

andrewhowdencom replied:
:heavy_check_mark:

andrewhowdencom committed:
AD-HOC refactor (stable/kolide-fleet): Code style, structure changes

This commit refactors some of the references that the chart uses such
that they're more in the pattern of the existing charts committed to
this repository. This includes:

- Removing simple default values
- Ensuring all component variables use the simplest API possible
- Amending some docs to be clearer
- Amending some references to old API styles to their new API
  counterparts

andrewhowdencom reviewed (commented):
I used checkboxes to denote stuff as I've done it.

unguiculus commented on incubator/kolide-fleet/values.yaml:
```diff
@@ -0,0 +1,76 @@
+    tag: "latest"
+    pullPolicy: "Always"
+  replicas: 1
+  pod:
```
I understand your concerns but it is commonly done as follows. Please change accordingly.

```yaml 
fleet:
  podAnnotations:
  resources: {}
```

unguiculus commented on incubator/kolide-fleet/values.yaml:
```diff
@@ -0,0 +1,76 @@
+fleet:
+  image:
+    repository: "kolide/fleet"
+    tag: "latest"
```
Use a stable image

unguiculus commented on incubator/kolide-fleet/templates/secret.yaml:
```diff
@@ -0,0 +1,15 @@
+data:
+  fleet.mysql.username: {{ required "A valid MySQL Username is required" .Values.mysql.mysqlUser | b64enc }}
+  fleet.mysql.password: {{ required "A valid MySQL Password is required" .Values.mysql.mysqlPassword | b64enc }}
+  fleet.redis.password: {{ required "A valid Redis password is required" .Values.redis.redisPassword | b64enc }}
```
This duplicates credentials which are already managed in secrets by their respective charts. Why don't you reuse these secrets?

andrewhowdencom replied:
So, when I was initially writing this I found it difficult to reconcile the credentials from the child chart to the parent chart over multiple releases (See commit https://github.com/kubernetes/charts/pull/5093/commits/d0e959db514158dcb54eb93150d6b98e0e29cfc9)

Essentially, because MySQL (absent a specific password supplied) would regenerate the password with reach release *but* not overwrite the password in the configuration, each `helm upgrade --install` would regenerate the password in the `helm` configuration (and in the MySQL configuration) but wouldn't apply to the mysql configuration. I think. It's been a while lol.

So, the first install would work fine, but subsequent installs would be unable to connect to the database.

If this should be possible, I can take another look at it, but my impression is that unless the MySQL password is stored somewhere in persistence when the application is booted then the upgrade of this chart will fail.

unguiculus commented on incubator/kolide-fleet/values.yaml:
```diff
@@ -0,0 +1,76 @@
+  logging:
+    debug: false
+    json: true
+    disable_banner: true
```
Add newline at end of file
